sdks/go/pkg/beam/io/mongodbio/read.go

```diff
@@ -442,53 +266,53 @@ func (fn *readFn) ProcessElement(
 	}()
 
 	for cursor.Next(ctx) {
-		value, err := decodeDocument(cursor, fn.Type.T)
+		id, value, err := decodeDocument(cursor, fn.Type.T)
 		if err != nil {
 			return err
 		}
 
-		emit(value)
-	}
-
-	return cursor.Err()
-}
+		result := cursorResult{nextID: id}
+		if !rt.TryClaim(result) {
+			return cursor.Err()
+		}
```

Comment:
I was going to comment that this should be *before* the work of decoding, but since we appear to not know which ID we're claiming until after decoding, this is fine.

Essentially, work should happen after a successful claim, but we're stuck with the API's we're given, and working non-ideal code is better than non-existent code that does what we need.